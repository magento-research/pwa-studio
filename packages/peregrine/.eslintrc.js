const config = {
    parser: 'babel-eslint',
    extends: ['@magento'],
    rules: {
        "no-undef": "off"
    }
};

module.exports = config;
