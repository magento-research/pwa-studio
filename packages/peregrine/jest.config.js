module.exports = {
    displayName: 'Peregrine',
    setupFiles: ['<rootDir>/scripts/shim.js', '<rootDir>/scripts/fetch-mock.js']
};
