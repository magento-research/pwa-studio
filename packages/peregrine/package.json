{
  "name": "@magento/peregrine",
  "version": "3.0.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "build": " ",
    "build:ci": " ",
    "build:cjs": " ",
    "build:dev": " ",
    "clean": " ",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "storybook:build": "build-storybook -c .storybook -o storybook-dist",
    "watch": " "
  },
  "repository": "github:magento-research/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento-research/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine#readme",
  "devDependencies": {
    "@storybook/react": "~4.1.13",
    "intl": "~1.2.5",
    "react": "~16.8.3",
    "react-apollo": "~2.5.3",
    "react-dom": "~16.8.3",
    "react-router-dom": "~5.0.0",
    "react-test-renderer": "~16.8.6",
    "rimraf": "~2.6.3"
  },
  "peerDependencies": {
    "@babel/runtime": "^7.4.2",
    "react": "^16.8.0",
    "react-apollo": "^2.5.3",
    "react-dom": "^16.8.0",
    "react-router-dom": "~5.0.0"
  },
  "engines": {
    "node": "8.10.0 || >=10.7.0",
    "yarn": ">=1.12.0"
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
