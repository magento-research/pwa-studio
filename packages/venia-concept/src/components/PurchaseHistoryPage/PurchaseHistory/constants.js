import PropTypes from 'prop-types';

export const PURCHASE_HISTORY_ITEM_PROP_TYPES = {
    id: PropTypes.number,
    imageSrc: PropTypes.string,
    title: PropTypes.string,
    date: PropTypes.date,
    link: PropTypes.string
};
