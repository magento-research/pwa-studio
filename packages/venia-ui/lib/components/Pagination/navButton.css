.root {
    composes: button from './tile.css';
    display: grid;
    padding-bottom: 1px;
}

.root_disabled {
    composes: root;
}

.icon {
    composes: root from '../Icon/icon.css';
}

.icon_disabled {
    composes: icon;
    --stroke: rgb(var(--venia-text-hint));
}
