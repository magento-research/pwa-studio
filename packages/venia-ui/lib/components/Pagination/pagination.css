.root {
    display: grid;
    grid-auto-columns: minmax(2rem, auto);
    grid-template-rows: 1fr;
    grid-column-gap: 0.25rem;
    justify-content: center;
    padding-top: 1rem;
    padding-bottom: 1rem;
    background-color: white;
    border-top: 1px solid rgb(var(--venia-border));
    text-align: center;
}
