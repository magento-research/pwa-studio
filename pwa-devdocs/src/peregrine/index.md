---
title: Peregrine
---

The Peregrine project is a collection of UI Components for Magento PWA projects.
Use, extend, or remix these components to create a unique Magento PWA storefront.

Visit the [`Peregrine`] repository to contribute to this project.

[`Peregrine`]: https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine
