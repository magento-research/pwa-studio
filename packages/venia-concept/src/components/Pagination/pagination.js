import React, { Component } from 'react';
import { func, number, object, shape, string } from 'prop-types';
import { withRouter } from 'src/drivers';
import { compose } from 'redux';

import classify from 'src/classify';
import defaultClasses from './pagination.css';
import NavButton from './navButton';
import { navButtons } from './constants';

const tileBuffer = 2;

class Pagination extends Component {
    static propTypes = {
        classes: shape({
            root: string
        }),
        history: object,
        location: object,
        pageControl: shape({
            currentPage: number,
            setPage: func,
            totalPages: number,
            updateTotalPages: func
        })
    };

    get navigationTiles() {
        const { classes, pageControl } = this.props;
        const { currentPage, setPage, totalPages } = pageControl;

        // Begin building page navigation tiles
        const tiles = [];
        const visibleBuffer = Math.min(tileBuffer * 2, totalPages - 1);
        const leadTile = this.getLeadTile(currentPage, totalPages);

        for (let i = leadTile; i <= leadTile + visibleBuffer; i++) {
            const tile = i;
            tiles.push(tile);
        }
        // End building page navigation tiles

        return tiles.map(tile => {
            const tileMarker =
                tile == currentPage ? (
                    <div className={classes.tileMarker} />
                ) : null;
            return (
                <button
                    className={classes.tileButton}
                    key={tile}
                    onClick={() => setPage(tile)}
                >
                    {tileMarker}
                    {tile}
                </button>
            );
        });
    }

    render() {
        const { classes } = this.props;
        const { currentPage, setPage, totalPages } = this.props.pageControl;
        const { navigationTiles } = this;

        if (!this.props.pageControl || totalPages == 1) {
            return null;
        }

        const leadTile = this.getLeadTile(currentPage, totalPages);

        const rightSkip = Math.min(
            totalPages,
            leadTile + tileBuffer * 2 + (tileBuffer + 1)
        );
        const leftSkip = Math.max(1, leadTile - (tileBuffer + 1));

        const isActiveLeft = !(currentPage == 1);
        const isActiveRight = !(currentPage == totalPages);

        return (
            <div className={classes.root}>
                <NavButton
                    name={navButtons.firstPage.name}
                    active={isActiveLeft}
                    onClick={() => setPage(leftSkip)}
                    buttonLabel={navButtons.firstPage.buttonLabel}
                />
                <NavButton
                    name={navButtons.prevPage.name}
                    active={isActiveLeft}
                    onClick={this.slideNavLeft}
                    buttonLabel={navButtons.prevPage.buttonLabel}
                />
                {navigationTiles}
                <NavButton
                    name={navButtons.nextPage.name}
                    active={isActiveRight}
                    onClick={this.slideNavRight}
                    buttonLabel={navButtons.nextPage.buttonLabel}
                />
                <NavButton
                    name={navButtons.lastPage.name}
                    active={isActiveRight}
                    onClick={() => setPage(rightSkip)}
                    buttonLabel={navButtons.lastPage.buttonLabel}
                />
            </div>
        );
    }

    slideNavLeft = () => {
        const { currentPage, setPage } = this.props.pageControl;
        if (currentPage > 1) {
            setPage(currentPage - 1);
        }
    };

    slideNavRight = () => {
        const { currentPage, setPage, totalPages } = this.props.pageControl;
        if (currentPage < totalPages) {
            setPage(currentPage + 1);
        }
    };

    getLeadTile = (currentPage, totalPages) => {
        const selectedTile = currentPage;
        const leftBound = 1 + tileBuffer;
        const rightBound = totalPages - tileBuffer;

        let leadTile = selectedTile - tileBuffer;
        if (selectedTile < leftBound) {
            leadTile = 1;
        } else if (selectedTile > rightBound) {
            leadTile = Math.max(totalPages - tileBuffer * 2, 1);
        }
        return leadTile;
    };
}

export default compose(
    withRouter,
    classify(defaultClasses)
)(Pagination);
