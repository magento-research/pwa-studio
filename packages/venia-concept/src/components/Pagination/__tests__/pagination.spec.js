import React from 'react';
import { act } from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { createTestInstance, usePagination } from '@magento/peregrine';

import { navButtons } from '../constants';
import NavButton from '../navButton';
import Pagination from '../pagination';

jest.mock('src/classify');
jest.mock('../navButton');
void NavButton;

const PaginationWrapper = ({ currentPage = 1, totalPages = 3 }) => {
    const [paginationValues, paginationApi] = usePagination(currentPage);

    const pageControl = {
        currentPage: paginationValues.currentPage || currentPage,
        setPage: jest.spyOn(paginationApi, 'setCurrentPage'),
        updateTotalPages: jest.spyOn(paginationApi, 'setTotalPages'),
        totalPages: paginationValues.totalPages || totalPages
    };
    return <Pagination pageControl={pageControl} />;
};

test('renders when there is more than 1 page', () => {
    const { root } = createTestInstance(
        <MemoryRouter>
            <PaginationWrapper />
        </MemoryRouter>
    );

    expect(root.findByProps({ className: 'root' })).toBeTruthy();
});

test('renders nothing when there is only 1 page', () => {
    const { root } = createTestInstance(
        <MemoryRouter>
            <PaginationWrapper totalPages={1} />
        </MemoryRouter>
    );

    expect(root.findAllByProps({ className: 'root' })).toHaveLength(0);
});

test('tiles set the appropriate page number on click', () => {
    const { root } = createTestInstance(
        <MemoryRouter>
            <PaginationWrapper />
        </MemoryRouter>
    );

    const tile = root.findAllByProps({ className: 'tileButton' })[2];

    act(() => {
        tile.props.onClick();
    });

    const setPageSpy = root.findByType(Pagination).props.pageControl.setPage;
    expect(setPageSpy).toHaveBeenCalledTimes(1);
    expect(setPageSpy).toHaveBeenLastCalledWith(3);
});

test('sets the current page based on the query param', async () => {
    const { root } = createTestInstance(
        <MemoryRouter
            initialEntries={[{ pathname: '/index.html', search: '?page=2' }]}
            initialIndex={0}
        >
            <PaginationWrapper />
        </MemoryRouter>
    );

    const setPageSpy = root.findByType(Pagination).props.pageControl.setPage;
    expect(setPageSpy).toHaveBeenCalledTimes(1);
    expect(setPageSpy).toHaveBeenLastCalledWith(2);
});

test('prevPage button decrements page by 1', () => {
    const { root } = createTestInstance(
        <MemoryRouter
            initialEntries={[{ pathname: '/index.html', search: '?page=2' }]}
            initialIndex={0}
        >
            <PaginationWrapper currentPage={2} />
        </MemoryRouter>
    );

    const prevPage = root.findByProps({ name: navButtons.prevPage.name });
    act(() => {
        prevPage.props.onClick();
    });

    const setPageSpy = root.findByType(Pagination).props.pageControl.setPage;
    expect(setPageSpy).toHaveBeenCalledTimes(1);
    expect(setPageSpy).toHaveBeenLastCalledWith(1);
});

test('prevPage button does nothing if currentPage is 1', () => {
    const { root } = createTestInstance(
        <MemoryRouter>
            <PaginationWrapper />
        </MemoryRouter>
    );

    const prevPage = root.findByProps({ name: navButtons.prevPage.name });

    act(() => {
        prevPage.props.onClick();
    });

    const setPageSpy = root.findByType(Pagination).props.pageControl.setPage;
    expect(setPageSpy).not.toHaveBeenCalled();
});

test('nextPage button increments page by 1', () => {
    const { root } = createTestInstance(
        <MemoryRouter>
            <PaginationWrapper />
        </MemoryRouter>
    );

    const nextPage = root.findByProps({ name: navButtons.nextPage.name });

    act(() => {
        nextPage.props.onClick();
    });

    const setPageSpy = root.findByType(Pagination).props.pageControl.setPage;
    expect(setPageSpy).toHaveBeenCalledTimes(1);
    expect(setPageSpy).toHaveBeenLastCalledWith(2);
});

test('nextPage button does nothing on last page', () => {
    const { root } = createTestInstance(
        <MemoryRouter
            initialEntries={[{ pathname: '/index.html', search: '?page=3' }]}
            initialIndex={0}
        >
            <PaginationWrapper currentPage={3} />
        </MemoryRouter>
    );

    const nextPage = root.findByProps({ name: navButtons.nextPage.name });

    act(() => {
        nextPage.props.onClick();
    });

    const setPageSpy = root.findByType(Pagination).props.pageControl.setPage;
    expect(setPageSpy).not.toHaveBeenCalled();
});

test('leftSkip skips toward first page', () => {
    const { root } = createTestInstance(
        <MemoryRouter
            initialEntries={[{ pathname: '/index.html', search: '?page=8' }]}
            initialIndex={0}
        >
            <PaginationWrapper currentPage={8} totalPages={10} />
        </MemoryRouter>
    );

    const firstPage = root.findByProps({ name: navButtons.firstPage.name });

    act(() => {
        firstPage.props.onClick();
    });

    const setPageSpy = root.findByType(Pagination).props.pageControl.setPage;
    expect(setPageSpy).toHaveBeenCalledTimes(1);
    expect(setPageSpy).toHaveBeenLastCalledWith(3);
});

test('rightSkip skips toward last page', () => {
    const { root } = createTestInstance(
        <MemoryRouter>
            <PaginationWrapper />
        </MemoryRouter>
    );

    const lastPage = root.findByProps({ name: navButtons.lastPage.name });
    const setPageSpy = root.findByType(Pagination).props.pageControl.setPage;

    act(() => {
        lastPage.props.onClick();
    });

    expect(setPageSpy).toHaveBeenCalledTimes(1);
    expect(setPageSpy).toHaveBeenLastCalledWith(3);
});
