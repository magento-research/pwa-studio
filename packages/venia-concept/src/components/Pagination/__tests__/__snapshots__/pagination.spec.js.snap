// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders nothing when there is only 1 page 1`] = `null`;

exports[`renders when there is more than 1 page 1`] = `
<div
  className="root"
>
  <button
    className="tileButton"
    onClick={[Function]}
  >
    <div
      className="tileMarker"
    />
    1
  </button>
  <button
    className="tileButton"
    onClick={[Function]}
  >
    2
  </button>
  <button
    className="tileButton"
    onClick={[Function]}
  >
    3
  </button>
</div>
`;
